proposals/references-to-java-synthetic-properties.md

```diff
@@ -0,0 +1,141 @@
+It also wasn't properly documented add covered with tests.
+Nonetheless, it was enabled by default in `.gradle.kts` (see [KT-35933](https://youtrack.jetbrains.com/issue/KT-35933)).
+Until recently, it didn't work in K2 (see [KT-54770](https://youtrack.jetbrains.com/issue/KT-54770)); the issue has been
+solved in our prototype.
+
+Our review of the implementation added in 1.3.70 uncovered a number of issues:
+- Most of the features provided by `kotlin-reflect`, such as `KProperty::visibility` worked incorrectly.
+- Synthetic Java properties weren't included in `KClass.members`.
+- Synthetic Java properties can be overshadowed by "physical" declarations with the same name.
+
+We are going to consider the found issues and possible ways of solving them below.
+
+## Reflection support
+
+As was mentioned above, all features requiring `kotlin-reflect`, such as `KProperty::visibility` worked incorrectly. 
+Instead of somehow taking into account the synthetic nature of the property, they tried to access a Java **field** of the same name. 
+If no such field existed a run-time exception would be thrown.
+In particular, that meant that calling `propertyReference.seter(value)` would bypass the original setter method and write
+directly to the underlying field!
+
+Possibly, a proper solution to this issue would be to implement proper reflection support.
+However, the demand for such a feature is most likely very low, so that we could postpone its implementation.
+Supporting reflection would require making some extra design decisions, for example on what should be returned
+by such properties as `KProperty::annotations` or `KProperty.Getter::name`.
+
+As a midterm solution, we propose to forbid `kotlin-reflect`-based reflection for synthetic Java properties.
+Any invocation of a method or a property requiring `kotlin-reflect` would result in an `UnsupportedOperationException`.
+That solution has been implemented in our prototype.
+If a significant number of users asks for full reflection, we'll reconsider this decision.
+
+The fact that synthetic Java properties aren't included in `KClass.members` doesn't seem to be an issue.
+Even though it wouldn't be difficult to include them, such a change could potentially break users' code.
+Very often Java synthetic properties use an underlying field whose name is the same as the name of the property itself.
+As we already include Java fields in `KClass.members`, adding also Java synthetic properties would mean that `KClass.members` would
+contain two `KProperty` entries with the same type and name for each synthetic property.
+User code using `KClass.members` to directly access Java fields may break after addition of extra members.
+This consideration seems to be a serious argument against inclusion of synthetic Java properties into `KClass.members`.
+```kotlin
+val nameField = Widget::class.memberProperties.single { it.name == "name"} // would fail if we include both the field and the synthetic property
```

Comment:
TBH the argument in this paragraph to keep this code working seems a bit weak, because it can fail anyway in case when you have member **extension** properties with the same name but different receiver types in the same class. It's possible also in a Java class, if they're inherited from a Kotlin supertype.

The "real" reason why they aren't included is that kotlin-reflect replicates the compiler logic, and from the compiler's point of view, they are not members, but extensions.

Not sure how to present it in the document though. Maybe it should just say something like: `members` only returns real members by design, so there are no synthetic properties.

Reply:
@udalov makes sense